docs/CollectionBestPractices.md

```diff
@@ -0,0 +1,360 @@
+- The frozen collections are not persistent; in fact, they can’t be mutated at all! Instead, frozen collections are
+  optimized for faster lookup operations — faster than their imperative counterparts.
+- Frozen collections provide faster lookup by performing up-front analysis and selecting an optimal implementation for
+  the content. This means that they are much more expensive to create.
+- Because of their higher creation cost and improved lookup performance, frozen collections are best suited for
+  static data.
+
+# Array Pools
+- When a temporary array is needed to perform work and the lifetime of the array is bounded, consider acquiring a
+  pooled array. `ArrayPool<T>` can be used to acquire an array of some minimum length that can be returned to the pool
+  when the work is done.
+
+> [!WARNING]
+> **Be mindful of the array size!**
+>
+> The size of an array acquired from an `ArrayPool<T>` is guaranteed to be at least as large as the minimum length that
+> was requested. However, it is likely that a larger array will have been returned. So, care should be taken to avoid
+> using the acquired array’s length, unless that’s what’s needed.
+
+- Razor provides a handful of helper extension methods that acquire pooled arrays and return them within the scope of a
+using statement:
+
+```C#
+var pool = ArrayPool<char>.Shared;
+
+using (pool.GetPooledArray(minimumLength: 42, out var array)
+{
+    // When using array but be careful that array.Length >= minimumLength.
+}
+
+using (pool.GetPooledArraySpan(minimumLength: 42, out var span)
+{
+    // span is array.AsSpan(0, minimumLength) to help avoid subtle bugs.
+}
+```
+
+# Object Pools
+- Razor provides object pooling facilities based on
+  [Microsoft.Extensions.ObjectPool](https://www.nuget.org/packages/Microsoft.Extensions.ObjectPool/) (which was
+  originally based on Roslyn’s `ObjectPool<T>`) along with several premade pools for many collection types in the
+  [Microsoft.AspNetCore.Razor.PooledObjects](https://github.com/dotnet/razor/tree/5c0677ad275e64300b897de0f6e8856ebe13f07b/src/Shared/Microsoft.AspNetCore.Razor.Utilities.Shared/PooledObjects)
+  namespace. These can be used to acquire temporary collections to use for work and return when finished.
+
+```C#
+using var _ = ListPool<int>.GetPooledObject(out var list);
+
+// Use list here. It'll be returned to the pool at the end of the using
+// statement's scope.
+```
+
+- Pooled collections provide a couple of benefits.
+  1. Pooled collections decrease pressure on the garbage collector by reusing collection instances.
+  2. Pooled collections avoid growing a collection’s internal storage. For example, when the `List<int>` acquired from
+     `ListPool<int>` in the code sample above is returned to the pool, it will be cleared. However, the capacity of its
+     internal storage will only be trimmed if it is larger than 512. So, lists acquired from the pool are likely to
+     already have a larger capacity than needed for most work.
+
```

Comment:
You've touched on it indirectly, but I think it would be worth being explicit about the downside of `ListPool` etc. that one should never return a pooled collection from a method, because of the clearning. ie, it really really is just for temporary work.

Reply:
I'll add a callout